libraries/adaptive-expressions/src/triggerTrees/trigger.ts

```diff
@@ -0,0 +1,502 @@
+                    const binding = quantifier.bindings[i];
+                    const newClause = new Clause(clause);
+                    const children: Expression[] = [];
+                    for (let j = 0; j < clause.children.length; j++) {
+                        const predicate = clause.children[j];
+                        const { expression: newPredicate, changed: predicateChanged } = this.substituteVariable(quantifier.variable, binding, predicate);
+                        changed = changed || predicateChanged;
+                        children.push(newPredicate);
+                    }
+
+                    if (changed) {
+                        newClause.anyBindings[quantifier.variable] = binding;
+                    }
+
+                    newClause.children = [...children];
+                    results.push(newClause);
+                    if (!changed) {
+                        break;
+                    }
+                }
+            } else {
+                // Keep clause if does not contain any binding
+                let changed = false;
+                for (let i = 0; i < clause.children.length; i++) {
+                    const predicate = clause.children[i];
+                    const { changed: predicateChanged } = this.substituteVariable(quantifier.variable, '', predicate);
+                    if (predicateChanged) {
+                        changed = true;
+                        break;
+                    }
+                }
+
+                if (!changed) {
+                    results.push(clause);
+                }
+            }
+        }
+
+        return results;
+    }
+
+    private substituteVariable(variable: string, binding: string, expression: Expression): { expression: Expression, changed: boolean } {
+        let newExpr = expression;
+        let changed = false;
+        if (expression.type === ExpressionType.Accessor
+            && expression.children.length === 1
+            && expression.children[0] instanceof Constant
+            && (typeof (expression.children[0] as Constant).value) === 'string'
+            && (expression.children[0] as Constant).value === variable) {
+            newExpr = Expression.makeExpression(ExpressionType.Accessor, undefined, new Constant(binding));
+            changed = true;
+        } else {
+            const children: Expression[] = [];
+            for (let i = 0; i < expression.children.length; i++) {
+                const child = expression.children[i];
+                const { expression: childExpr, changed: childChanged } = this.substituteVariable(variable, binding, child);
+                children.push(childExpr);
+                changed = changed || childChanged;
+            }
+
+            if (changed) {
+                newExpr = new Expression(undefined, expression.evaluator, ...children);
+            }
+        }
+
+        return { expression: newExpr, changed };
+    }
+
+    private removeDuplicates(): void {
```

Comment:
It seems like use of underscore prefixes is not consistent - what is the convention or rule that we try to follow?

Reply:
I would use underscores for all private methods.